From foo@baz Tue Apr  9 12:12:43 2002
Date: Fri, 11 Aug 2006 01:55:12 -0700
From: Greg Kroah-Hartman <gregkh@suse.de>
Subject: USB: create a new thread for every USB device found during the probe sequence


Might speed up some systems.  If nothing else, a bad driver should not
take the whole USB subsystem down with it.

Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/usb/core/hub.c |   14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

--- gregkh-2.6.orig/drivers/usb/core/hub.c
+++ gregkh-2.6/drivers/usb/core/hub.c
@@ -1199,8 +1199,9 @@ static void show_string(struct usb_devic
  *
  * Only the hub driver or root-hub registrar should ever call this.
  */
-int usb_new_device(struct usb_device *udev)
+static int __usb_new_device(void *void_data)
 {
+	struct usb_device *udev = void_data;
 	int err;
 
 	err = usb_get_configuration(udev);
@@ -1311,6 +1312,17 @@ fail:
 	return err;
 }
 
+int usb_new_device(struct usb_device *udev)
+{
+	struct task_struct *probe_task;
+	int ret = 0;
+
+	probe_task = kthread_run(__usb_new_device, udev,
+				 "usb-probe-%s", udev->devnum);
+	if (IS_ERR(probe_task))
+		ret = PTR_ERR(probe_task);
+	return ret;
+}
 
 static int hub_port_status(struct usb_hub *hub, int port1,
 			       u16 *status, u16 *change)
